File: WordPig/app/src/main/java/com/pigdogbay/wordpig/MainActivity.java
package com.pigdogbay.wordpig;

import android.app.Activity;
import android.graphics.Canvas;
import android.graphics.Point;
import android.os.Bundle;
import android.view.Window;
import android.view.WindowManager;

import com.pigdogbay.lib.utils.ObservableProperty;
import com.pigdogbay.library.games.FrameBuffer;
import com.pigdogbay.library.games.GameView;
import com.pigdogbay.library.games.ObjectTouchHandler;
import com.pigdogbay.wordpig.model.Board;
import com.pigdogbay.wordpig.model.Screen;
import com.pigdogbay.wordpig.model.WordChecker;


public class MainActivity extends Activity implements GameView.IGame, ObservableProperty.PropertyChangedObserver<Screen.ScreenState> {

    private GameView gameView;
    private GameView.IGame currentScreen;
    private GameScreen gameScreen;
    private HomeScreen homeScreen;
    private GameOverScreen gameOverScreen;

    private FrameBuffer buffer;
    private ObjectTouchHandler touchHandler;
    private Screen screen;
    private Board board;
    private Assets assets;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //full screen
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        //Set up touch handler
        Point size = new Point();
        this.getWindowManager().getDefaultDisplay().getSize(size);
        touchHandler = new ObjectTouchHandler();
        float xScale = ((float)Defines.BOARD_WIDTH)/((float)size.x);
        float yScale = ((float)Defines.BOARD_HEIGHT)/((float)size.y);
        touchHandler.setXScale(xScale);
        touchHandler.setYScale(yScale);

        //Load game
        this.assets = new Assets();
        this.assets.load(this);
        createModel();
        createScreens();

        gameView = new GameView(this, this);
        gameView.setShowFPS(true);
        gameView.setOnTouchListener(touchHandler);
        setContentView(gameView);
        showHome();
    }

    @Override
    protected void onResume() {
        super.onResume();
        gameView.resume();
    }

    @Override
    protected void onPause() {
        super.onPause();
        gameView.pause();
    }

    private void createModel()
    {
        screen = new Screen();
        screen.screenStateObserver.addObserver(this);
    }

    private void createScreens()
    {
        buffer = new FrameBuffer(Defines.BOARD_WIDTH, Defines.BOARD_HEIGHT);
        gameScreen = new GameScreen();
        gameScreen.setTouchHandler(touchHandler);
        gameScreen.setBuffer(buffer);
        gameScreen.setAssets(this.assets);
        gameScreen.initialize();

        homeScreen = new HomeScreen();
        homeScreen.setTouchHandler(touchHandler);
        homeScreen.setBuffer(buffer);
        homeScreen.setScreen(screen);
        homeScreen.setAssets(this.assets);
        homeScreen.initialize();

        gameOverScreen = new GameOverScreen();
        gameOverScreen.setTouchHandler(touchHandler);
        gameOverScreen.setBuffer(buffer);
        gameOverScreen.setScreen(screen);
        gameOverScreen.setAssets(this.assets);
        gameOverScreen.initializie();
    }

    private void showGame()
    {
        board = new Board(this.screen);
        board.setTiles("streaming");
        board.wordChecker = new WordChecker(this.assets.wordList);
        gameScreen.setBoard(board);
        touchHandler.clearTouchables();
        gameScreen.registerTouchables();
        currentScreen = gameScreen;
    }

    private void showHome()
    {
        touchHandler.clearTouchables();
        homeScreen.registerTouchables();
        currentScreen = homeScreen;
    }
    private void showGameOver()
    {
        touchHandler.clearTouchables();
        gameOverScreen.registerTouchables();
        currentScreen = gameOverScreen;
    }

    //IGame Implementation
    @Override
    public void Update() {
        currentScreen.Update();
    }

    @Override
    public void Render(Canvas c) {
        currentScreen.Render(c);
    }

    //Game state changes
    @Override
    public void update(ObservableProperty<Screen.ScreenState> sender, Screen.ScreenState update) {
        switch (update){
            case Home:
                showHome();
                break;
            case Game:
                showGame();
                break;
            case GameOver:
                showGameOver();
                break;
        }

    }
}

File: WordPig/app/src/main/java/com/pigdogbay/wordpig/MainActivity.kt
package com.pigdogbay.wordpig

import android.annotation.SuppressLint
import android.app.Activity
import android.graphics.Canvas
import android.graphics.Point
import android.os.Build
import android.os.Bundle
import android.view.Window
import android.view.WindowInsets
import android.view.WindowInsetsController
import android.view.WindowManager
import com.pigdogbay.lib.games.GameView
import com.pigdogbay.lib.games.GameView.Game
import com.pigdogbay.lib.utils.ObservableProperty
import com.pigdogbay.wordpig.model.ScreenState
import com.pigdogbay.wordpig.model.WordChecker

class MainActivity : Activity(), Game, ObservableProperty.PropertyChangedObserver<ScreenState> {
    private lateinit var gameView: GameView
    private lateinit var currentScreen: Game
    private val gameScreen: GameScreen
        get() = Injector.gameScreen
    private val homeScreen: HomeScreen
        get() = Injector.homeScreen
    private val gameOverScreen: GameOverScreen
        get() = Injector.gameOverScreen

    @Suppress("DEPRECATION")
    @SuppressLint("ClickableViewAccessibility")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R){
            fullScreen()
        }

        //Set up touch handler
        val size = Point()
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R){
            //On the Samsung this version gives 2132x1080
            this.windowManager.defaultDisplay.getSize(size)
        } else {
            //On the Samsung this version gives 2340x1080
            val bounds = this.windowManager.currentWindowMetrics.bounds
            size.x = bounds.width()
            size.y = bounds.height()
        }
        Assets.load(this.applicationContext)
        Injector.build(size)
        Injector.createScreens()
        gameView = GameView(this, this)
        gameView.setShowFPS(true)
        gameView.setOnTouchListener(Injector.touchHandler)
        setContentView(gameView)
        showHome()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        //Full screen for Android R+
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R){
            window.insetsController?.hide(WindowInsets.Type.statusBars() or WindowInsets.Type.navigationBars())
            window.insetsController?.systemBarsBehavior = WindowInsetsController.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        }
    }

    @Suppress("DEPRECATION")
    private fun fullScreen() {
        //full screen
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        window.setFlags(
            WindowManager.LayoutParams.FLAG_FULLSCREEN,
            WindowManager.LayoutParams.FLAG_FULLSCREEN
        )
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    override fun onResume() {
        super.onResume()
        gameView.resume()
        Injector.model.screenStateObserver.addObserver(this)
    }

    override fun onPause() {
        super.onPause()
        gameView.pause()
        Injector.model.screenStateObserver.removeObserver(this)
    }

    private fun showGame() {
        val board = Injector.board
        board.setTiles("streaming")
        board.wordChecker = WordChecker(Assets.wordList)
        Injector.touchHandler.clearTouchables()
        gameScreen.registerTouchables()
        currentScreen = gameScreen
    }

    private fun showHome() {
        Injector.touchHandler.clearTouchables()
        homeScreen.registerTouchables()
        currentScreen = homeScreen
    }

    private fun showGameOver() {
        Injector.touchHandler.clearTouchables()
        gameOverScreen.registerTouchables()
        currentScreen = gameOverScreen
    }

    //Game Implementation
    override fun update() {
        currentScreen.update()
    }

    override fun render(c: Canvas?) {
        currentScreen.render(c)
    }

    //Game state changes
    override fun update(sender: ObservableProperty<ScreenState?>?, update: ScreenState?) {
        when (update!!) {
            ScreenState.Home -> showHome()
            ScreenState.Game -> showGame()
            ScreenState.GameOver -> showGameOver()
        }
    }
}
